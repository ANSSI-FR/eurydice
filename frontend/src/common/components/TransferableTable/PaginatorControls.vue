<template>
  <div class="Paginator-controls mt-3" :style="cssProps">
    <v-btn
      outlined
      :disabled="page === 1 || disabled"
      color="grey darken-2"
      class="ml-2"
      @click="previous"
    >
      <v-icon>{{ mdiChevronLeft }}</v-icon>
    </v-btn>
    <v-text-field
      :value="page"
      label="Page"
      :suffix="suffixString"
      dense
      outlined
      :disabled="disabled"
      class="shrink mx-2"
      :rules="rules"
      @keyup.enter="goToPage"
    />
    <v-btn
      outlined
      :disabled="page === numPages || disabled"
      color="grey darken-2"
      class="mr-4"
      @click="next"
    >
      <v-icon>{{ mdiChevronRight }}</v-icon>
    </v-btn>
  </div>
</template>

<script>
import { mdiChevronLeft, mdiChevronRight } from "@mdi/js";

export default {
  name: "PaginatorControls",
  props: {
    numPages: {
      type: Number,
      required: true,
      default: 0,
    },
    loading: {
      type: Boolean,
      required: false,
      default: false,
    },
  },
  data() {
    return {
      page: 1,
      mdiChevronLeft,
      mdiChevronRight,
    };
  },
  computed: {
    suffixString() {
      return `sur ${this.numPages}`;
    },
    rules() {
      return [
        (v) => (v && v >= 1) || "Doit être supérieur à 0",
        (v) =>
          (v && v <= this.numPages) || `Doit être inférieur à ${this.numPages}`,
      ];
    },
    /**
     * Compute the text input CSS width from the largest string that could credibly be inputted.
     * @return  {Object} val  The CSS computed CSS props as an object.
     */
    cssProps() {
      return {
        "--text-field-width": `${
          this.numPages.toString().length + this.suffixString.length + 3
        }ch`,
      };
    },
    disabled() {
      return this.loading || this.numPages === 0;
    },
  },
  methods: {
    async next() {
      this.page += 1;
      this.$emit("next");
    },
    async previous() {
      this.page -= 1;
      this.$emit("previous");
    },
    goToPage(event) {
      const pageNumber = parseInt(event.target.value, 10);
      if (pageNumber && pageNumber >= 1 && pageNumber <= this.numPages) {
        this.page = pageNumber;
        this.$emit("goto-page", this.page);
      }
    },
  },
};
</script>

<style lang="scss" scoped>
.v-btn:not(.v-btn--round).v-size--default {
  height: 40px;
  min-width: 0;
  width: 10px;
}

.v-text-field.v-text-field--enclosed {
  // input validation messages take up more than one line if smaller
  min-width: 16ch;
  // just the right size to allow inputting the largest page number without any overflow
  width: var(--text-field-width);
}

.Paginator-controls {
  align-items: flex-start;
  display: flex;
  justify-content: flex-end;
}
</style>
