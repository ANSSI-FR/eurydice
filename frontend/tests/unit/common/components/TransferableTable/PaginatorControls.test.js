import PaginatorControls from "@common/components/TransferableTable/PaginatorControls";
import { createLocalVue, mount, shallowMount } from "@vue/test-utils";
import Vuetify from "vuetify";

describe("PaginatorControls", () => {
  it("renders correctly", async () => {
    const localVue = createLocalVue();
    const vuetify = new Vuetify();
    const wrapper = mount(PaginatorControls, {
      localVue,
      vuetify,
      propsData: {
        numPages: 6,
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it.each`
    requestedPage                         | result
    ${Number.MIN_SAFE_INTEGER.toString()} | ${1}
    ${"-1"}                               | ${1}
    ${"0"}                                | ${1}
    ${"1"}                                | ${1}
    ${"5"}                                | ${5}
    ${"10"}                               | ${10}
    ${"11"}                               | ${1}
    ${"a"}                                | ${1}
    ${""}                                 | ${1}
    ${Number.MAX_SAFE_INTEGER.toString()} | ${1}
  `(
    `goToPage should set the current page to $result when the page requested is $requestedPage`,
    ({ requestedPage, result }) => {
      const wrapper = shallowMount(PaginatorControls, {
        attrs: {
          page: 1,
          numPages: 10,
        },
      });

      const event = {
        target: { value: requestedPage },
      };

      wrapper.vm.goToPage(event);

      expect(wrapper.vm.page).toBe(result);
    }
  );
});
