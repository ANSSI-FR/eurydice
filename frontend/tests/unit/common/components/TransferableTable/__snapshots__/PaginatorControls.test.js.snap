// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginatorControls renders correctly 1`] = `
<div
  class="Paginator-controls mt-3"
  style="--text-field-width: 9ch;"
>
  <button
    class="ml-2 v-btn v-btn--disabled v-btn--outlined theme--light v-size--default"
    disabled="disabled"
    type="button"
  >
    <span
      class="v-btn__content"
    >
      <span
        aria-hidden="true"
        class="v-icon notranslate theme--light"
      >
        <svg
          aria-hidden="true"
          class="v-icon__svg"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.41,16.58L10.83,12L15.41,7.41L14,6L8,12L14,18L15.41,16.58Z"
          />
        </svg>
      </span>
    </span>
  </button>
   
  <div
    class="v-input shrink mx-2 v-input--is-label-active v-input--is-dirty v-input--dense theme--light v-text-field v-text-field--enclosed v-text-field--outlined"
  >
    <div
      class="v-input__control"
    >
      <div
        class="v-input__slot"
      >
        <fieldset
          aria-hidden="true"
        >
          <legend
            style="width: 0px;"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
        <div
          class="v-text-field__slot"
        >
          <label
            class="v-label v-label--active theme--light"
            for="input-4"
            style="left: 0px; position: absolute;"
          >
            Page
          </label>
          <input
            id="input-4"
            type="text"
          />
          <div
            class="v-text-field__suffix"
          >
            sur 6
          </div>
        </div>
      </div>
      <div
        class="v-text-field__details"
      >
        <div
          class="v-messages theme--light"
        >
          <transition-group-stub
            class="v-messages__wrapper"
            name="message-transition"
            tag="div"
          />
        </div>
      </div>
    </div>
  </div>
   
  <button
    class="mr-4 v-btn v-btn--outlined theme--light v-size--default grey--text text--darken-2"
    type="button"
  >
    <span
      class="v-btn__content"
    >
      <span
        aria-hidden="true"
        class="v-icon notranslate theme--light"
      >
        <svg
          aria-hidden="true"
          class="v-icon__svg"
          role="img"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8.59,16.58L13.17,12L8.59,7.41L10,6L16,12L10,18L8.59,16.58Z"
          />
        </svg>
      </span>
    </span>
  </button>
</div>
`;
